File: analysis/revision/scripts/absolute.py
import pandas as pd, numpy as np
from scipy import stats
from sklearn.mixture import GaussianMixture
import matplotlib; matplotlib.use('Agg'); import matplotlib.pyplot as plt

np.random.seed(0)
df=pd.read_csv('./data/Tree_ID_info.csv')
df['site']=np.where(df['plot']=='BGS','BGS','EMS')
sppmap={'rm':'A.rubrum','bg':'N.sylvatica','ro':'Q.rubra','hem':'T.canadensis'}
df['sp']=df['species'].map(sppmap)
ert=pd.read_csv('./data/ERT_application_results.csv')
m=df.merge(ert,on='tree',how='inner').reset_index(drop=True)
m['res']=m['mean']            # mean resistivity, Ohm-m (absolute)
m['struct']=m['percent_damaged']>1
ns=(~m['struct']).values
mb=(m.site=='BGS').values; me=(m.site=='EMS').values

# ---- validation calibration: moisture ~ resistivity (n=12) ----
v=pd.read_csv('./data/hemlock/validation_summary.csv')
v['res']=1000/v['Conductance']
sl=stats.linregress(v['res'],v['moisture'])
n=len(v); xbar=v['res'].mean(); Sxx=((v['res']-xbar)**2).sum()
resid=v['moisture']-(sl.intercept+sl.slope*v['res']); rstd=np.sqrt((resid**2).sum()/(n-2))
def pred_moist(x): return sl.intercept+sl.slope*np.asarray(x,float)
def pred_se(x):    return rstd*np.sqrt(1+1/n+(np.asarray(x,float)-xbar)**2/Sxx)   # prediction SE
def res_for_moist(Mcut):  # invert: resistivity giving predicted moisture=Mcut, with band
    x=(Mcut-sl.intercept)/sl.slope
    # approximate inverse band via slope
    band=pred_se(x)/abs(sl.slope)
    return x, x-1.96*band, x+1.96*band
m['pmoist']=pred_moist(m['res'])
r100=res_for_moist(100)[0]; print(f"Calibration: moisture(%) = {sl.intercept:.1f} {sl.slope:+.4f}*res   r={sl.rvalue:.2f}, n={n}, resid s={rstd:.1f}%")
print(f"Predicted moisture range across 57 trees: {m['pmoist'].min():.0f}-{m['pmoist'].max():.0f}%")

print("\n--- Absolute mean resistivity (Ohm-m) by species & site ---")
print(m.groupby('sp')['res'].agg(['count','mean','min','max']).round(0))
print(m.groupby('site')['res'].agg(['mean','median']).round(0).to_dict())

# ---- modality on absolute resistivity ----
def modality(x,label):
    x=np.asarray(x,float).reshape(-1,1)
    b1=GaussianMixture(1,random_state=0).fit(x).bic(x)
    b2=GaussianMixture(2,random_state=0,n_init=5).fit(x).bic(x)
    print(f"{label}: BIC 1={b1:.1f} 2={b2:.1f} -> {'bimodal' if b2+2<b1 else 'unimodal'}; skew={stats.skew(x.ravel()):.2f}")
modality(m['res'],'abs resistivity')
modality(m['pmoist'],'pred moisture')

# ---- candidate ABSOLUTE thresholds (anomaly = LOW resistivity / HIGH moisture) ----
def otsu(x,nb=256,invert=False):
    h,e=np.histogram(x,bins=nb); c=(e[:-1]+e[1:])/2; w=h/h.sum()
    cw=np.cumsum(w); cm=np.cumsum(w*c); gm=cm[-1]; bt,bv=None,-1
    for i in range(1,nb):
        w0=cw[i]; w1=1-w0
        if w0<=0 or w1<=0: continue
        val=w0*w1*((cm[i]/w0)-((gm-cm[i])/w1))**2
        if val>bv: bv,bt=val,c[i]
    return bt
res=m['res'].values
t_otsu=otsu(res)
g2=GaussianMixture(2,random_state=0,n_init=5).fit(res.reshape(-1,1))
mns=np.sort(g2.means_.ravel()); xs=np.linspace(mns[0],mns[1],2000)
o=np.argsort(g2.means_.ravel()); lp=g2._estimate_weighted_log_prob(xs.reshape(-1,1))
dd=lp[:,o[0]]-lp[:,o[1]]; ix=np.where(np.diff(np.sign(dd))!=0)[0]; t_gmm=xs[ix[0]] if len(ix) else np.nan

# absolute anomaly = res < threshold
rules=[
 ('Sample median resistivity', np.median(res)),
 ('Sample mean resistivity', res.mean()),
 ('Otsu break (data-driven)', t_otsu),
 ('GMM crossover (data-driven)', t_gmm),
 ('Lit: silver-fir wetwood ~200', 200.0),
 ('Lit: silver-fir low-ER ~166', 166.0),
 ('Validation: moisture>90% (<505)', res_for_moist(90)[0]),
 ('Validation: moisture>100% (<402)', res_for_moist(100)[0]),
 ('Validation: moisture>110% (<299)', res_for_moist(110)[0]),
]
print("\n"+"="*92)
print(f"{'ABSOLUTE threshold rule':<34}{'res<':>7}{'~moist>':>9}{'nAnom':>7}{'nIncip':>7}{'BGSinc%':>9}{'EMSinc%':>9}")
print("="*92)
tbl=[]
for name,t in rules:
    anom=res<t
    inc=ns&anom
    bgs=inc[mb].sum()/mb.sum()*100; ems=inc[me].sum()/me.sum()*100
    mo=pred_moist(t)
    print(f"{name:<34}{t:>7.0f}{mo:>9.0f}{anom.sum():>7}{inc.sum():>7}{bgs:>9.0f}{ems:>9.0f}")
    tbl.append(dict(rule=name,res_thresh=round(t),moist_equiv=round(mo),n_anom=int(anom.sum()),
                    n_incip=int(inc.sum()),BGS_inc=round(bgs),EMS_inc=round(ems)))
pd.DataFrame(tbl).to_csv('analysis/revision/output/CJFR-absolute-thresholds.csv',index=False)

# stability of anomaly label across absolute rule set
lab=np.vstack([(res<t).astype(int) for _,t in rules])
print(f"\nAlways-anomalous (all abs rules): {(lab.min(0)==1).sum()}, never: {(lab.max(0)==0).sum()}, unstable: {(lab.min(0)!=lab.max(0)).sum()}")

# does site DIRECTION hold on absolute axis?
grid=np.linspace(np.percentile(res,5),np.percentile(res,95),200)
bgs=[]; ems=[]
for t in grid:
    inc=ns&(res<t); bgs.append(inc[mb].sum()/mb.sum()*100); ems.append(inc[me].sum()/me.sum()*100)
bgs=np.array(bgs); ems=np.array(ems)
print(f"\nSite DIRECTION on absolute axis: BGS>=EMS at {np.mean(bgs>=ems)*100:.0f}% of thresholds; "
      f"BGS>EMS at {np.mean(bgs>ems)*100:.0f}%; crossovers where EMS>BGS at {np.mean(ems>bgs)*100:.0f}%")
print(f"BGS incip range {bgs.min():.0f}-{bgs.max():.0f}%, EMS {ems.min():.0f}-{ems.max():.0f}%")

# ---------------- FIGURE (3 panels) ----------------
fig,axes=plt.subplots(1,3,figsize=(16.5,4.9))
spp_order=['A.rubrum','T.canadensis','N.sylvatica','Q.rubra']
cols={'A.rubrum':'#2c7fb8','T.canadensis':'#41b6c4','N.sylvatica':'#7fcdbb','Q.rubra':'#d95f0e'}

# A: absolute resistivity by species (the confound made visible)
ax=axes[0]
for i,s in enumerate(spp_order):
    d=m.loc[m.sp==s,'res']
    ax.scatter(d, np.random.normal(i,0.08,len(d)), color=cols[s], s=34, alpha=.8, edgecolor='white',lw=.5)
    ax.plot([d.mean(),d.mean()],[i-.28,i+.28],color=cols[s],lw=2.5)
ax.set_yticks(range(4)); ax.set_yticklabels([s+('*' if s in('N.sylvatica','Q.rubra') else '') for s in spp_order])
ax.set_xlabel('mean resistivity (Ω·m)  — lower = wetter'); ax.invert_xaxis()
ax.set_title('A.  Absolute axis is species-confounded\n(* = single-site specialist)',fontsize=10,loc='left')

# B: predicted-moisture distribution by site with candidate thresholds
ax=axes[1]
for s,msk,c in [('BGS',mb,'#1f77b4'),('EMS',me,'#d9822b')]:
    xs2=np.linspace(m['pmoist'].min()-5,m['pmoist'].max()+5,300)
    ax.plot(xs2,stats.gaussian_kde(m.loc[msk,'pmoist'])(xs2),color=c,lw=2,label=s)
    ax.scatter(m.loc[msk,'pmoist'],np.full(msk.sum(),-0.001)+ (0.0004 if s=='EMS' else 0),color=c,s=14,alpha=.6)
for Mc,c in [(90,'#999'),(100,'#555'),(110,'#111')]:
    ax.axvline(Mc,color=c,ls=':',lw=1.2); ax.text(Mc,ax.get_ylim()[1]*0.92,f'{Mc}%',fontsize=7,ha='center')
ax.set_xlabel('predicted moisture content (%)  [from validation calibration]'); ax.set_ylabel('density')
ax.legend(fontsize=9,frameon=False); ax.set_title('B.  Absolute (predicted moisture) axis by site',fontsize=10,loc='left')

# C: site direction sweep on absolute resistivity, anchors marked
ax=axes[2]
ax.plot(grid,bgs,color='#1f77b4',lw=2.4,label='BGS (wetland)')
ax.plot(grid,ems,color='#d9822b',lw=2.4,label='EMS (upland)')
ax.fill_between(grid,ems,bgs,where=bgs>=ems,color='#1f77b4',alpha=.10)
ax.fill_between(grid,ems,bgs,where=ems>bgs,color='#d9822b',alpha=.18)
anchor={'median':np.median(res),'mean':res.mean(),'Otsu':t_otsu,'moist>100%':r100,'wetwood~200':200}
for nm,t in anchor.items():
    ax.axvline(t,color='#444',ls=':',lw=1.0); ax.text(t,ax.get_ylim()[1]*0.98,nm,rotation=90,va='top',ha='right',fontsize=6.5,color='#444')
ax.set_xlabel('anomaly threshold: resistivity below … (Ω·m)'); ax.set_ylabel('% of site trees "incipient"')
ax.legend(fontsize=9,frameon=False,loc='upper left'); ax.invert_xaxis()
ax.set_title('C.  Site direction across absolute thresholds\n(shaded orange = where upland exceeds wetland)',fontsize=10,loc='left')
plt.tight_layout(); plt.savefig('analysis/revision/output/CJFR-absolute-fig.png',dpi=170)
print("\nsaved figure + CSV")

File: analysis/revision/scripts/absolute.R
# R port of absolute.py (revision sensitivity analysis, cjfr-2026-0202).
# Run from the repo root: Rscript analysis/revision/scripts/absolute.R
source("analysis/revision/scripts/revision_common.R")

skewness <- function(x) { mu <- mean(x); s <- sqrt(mean((x - mu)^2)); mean((x - mu)^3) / s^3 }

m <- load_merged()
m$res <- m$mean               # mean resistivity, Ohm-m (absolute)
m$struct <- m$percent_damaged > 1
ns <- !m$struct
mb <- m$site == "BGS"; me <- m$site == "EMS"

# ---- validation calibration: moisture ~ resistivity (n=12) ----
v <- load_validation()
v$res <- v$mean_res
fit <- lm(moisture ~ res, data = v)
itc <- unname(coef(fit)[1]); slp <- unname(coef(fit)[2])
rv <- cor(v$res, v$moisture)
n <- nrow(v); xbar <- mean(v$res); Sxx <- sum((v$res - xbar)^2)
resid <- v$moisture - (itc + slp * v$res)
rstd <- sqrt(sum(resid^2) / (n - 2))
pred_moist <- function(x) itc + slp * x
pred_se <- function(x) rstd * sqrt(1 + 1 / n + (x - xbar)^2 / Sxx)   # prediction SE
res_for_moist <- function(Mcut) {  # invert: resistivity giving predicted moisture=Mcut, with band
  x <- (Mcut - itc) / slp
  band <- pred_se(x) / abs(slp)
  c(x, x - 1.96 * band, x + 1.96 * band)
}
m$pmoist <- pred_moist(m$res)
r100 <- res_for_moist(100)[1]
cat(sprintf("Calibration: moisture(%%) = %.1f %+.4f*res   r=%.2f, n=%d, resid s=%.1f%%\n",
            itc, slp, rv, n, rstd))
cat(sprintf("Predicted moisture range across 57 trees: %.0f-%.0f%%\n",
            min(m$pmoist), max(m$pmoist)))

cat("\n--- Absolute mean resistivity (Ohm-m) by species & site ---\n")
spp <- sort(unique(m$sp))
cat(sprintf("%-13s%6s%7s%7s%7s\n", "", "count", "mean", "min", "max"))
cat(sprintf("%-13s\n", "sp"))
for (s in spp) {
  d <- m$res[m$sp == s]
  cat(sprintf("%-13s%6d%7.1f%7.1f%7.1f\n", s, length(d),
              round(mean(d)), round(min(d)), round(max(d))))
}
cat(sprintf("{'mean': {'BGS': %.1f, 'EMS': %.1f}, 'median': {'BGS': %.1f, 'EMS': %.1f}}\n",
            round(mean(m$res[mb])), round(mean(m$res[me])),
            round(median(m$res[mb])), round(median(m$res[me]))))

# ---- modality on absolute resistivity ----
modality <- function(x, label) {
  g <- gmm_1d(x)
  verdict <- if (g$bic2 + 2 < g$bic1) "bimodal" else "unimodal"
  cat(sprintf("%s: BIC 1=%.1f 2=%.1f -> %s; skew=%.2f\n",
              label, g$bic1, g$bic2, verdict, skewness(x)))
  invisible(g)
}
modality(m$res, "abs resistivity")
modality(m$pmoist, "pred moisture")

# ---- candidate ABSOLUTE thresholds (anomaly = LOW resistivity / HIGH moisture) ----
res <- m$res
t_otsu <- otsu_threshold(res)
t_gmm <- gmm_1d(res)$crossover
rules <- list(
  list(name = "Sample median resistivity",         t = median(res)),
  list(name = "Sample mean resistivity",           t = mean(res)),
  list(name = "Otsu break (data-driven)",          t = t_otsu),
  list(name = "GMM crossover (data-driven)",       t = t_gmm),
  list(name = "Lit: silver-fir wetwood ~200",      t = 200.0),
  list(name = "Lit: silver-fir low-ER ~166",       t = 166.0),
  list(name = "Validation: moisture>90% (<505)",   t = res_for_moist(90)[1]),
  list(name = "Validation: moisture>100% (<402)",  t = res_for_moist(100)[1]),
  list(name = "Validation: moisture>110% (<299)",  t = res_for_moist(110)[1]))
bar92 <- strrep("=", 92)
cat("\n", bar92, "\n", sep = "")
cat(sprintf("%-34s%7s%9s%7s%7s%9s%9s\n",
            "ABSOLUTE threshold rule", "res<", "~moist>", "nAnom", "nIncip", "BGSinc%", "EMSinc%"))
cat(bar92, "\n", sep = "")
tbl <- list()
for (r in rules) {
  anom <- res < r$t
  inc <- ns & anom
  bgsr <- sum(inc[mb]) / sum(mb) * 100; emsr <- sum(inc[me]) / sum(me) * 100
  mo <- pred_moist(r$t)
  cat(sprintf("%-34s%7.0f%9.0f%7d%7d%9.0f%9.0f\n",
              r$name, r$t, mo, sum(anom), sum(inc), bgsr, emsr))
  tbl[[length(tbl) + 1]] <- data.frame(
    rule = r$name, res_thresh = as.integer(round(r$t)),
    moist_equiv = as.integer(round(mo)), n_anom = sum(anom), n_incip = sum(inc),
    BGS_inc = as.integer(round(bgsr)), EMS_inc = as.integer(round(emsr)))
}
write.csv(do.call(rbind, tbl),
          file.path(OUT_DIR, "CJFR-absolute-thresholds.csv"),
          row.names = FALSE, quote = FALSE)

# stability of anomaly label across absolute rule set
lab <- sapply(rules, function(r) as.integer(res < r$t))   # 57 x 9
lab_min <- apply(lab, 1, min); lab_max <- apply(lab, 1, max)
cat(sprintf("\nAlways-anomalous (all abs rules): %d, never: %d, unstable: %d\n",
            sum(lab_min == 1), sum(lab_max == 0), sum(lab_min != lab_max)))

# does site DIRECTION hold on absolute axis?
grid <- seq(quantile(res, 0.05, type = 7), quantile(res, 0.95, type = 7), length.out = 200)
bgs <- numeric(length(grid)); ems <- numeric(length(grid))
for (i in seq_along(grid)) {
  inc <- ns & (res < grid[i])
  bgs[i] <- sum(inc[mb]) / sum(mb) * 100
  ems[i] <- sum(inc[me]) / sum(me) * 100
}
cat(sprintf("\nSite DIRECTION on absolute axis: BGS>=EMS at %.0f%% of thresholds; BGS>EMS at %.0f%%; crossovers where EMS>BGS at %.0f%%\n",
            mean(bgs >= ems) * 100, mean(bgs > ems) * 100, mean(ems > bgs) * 100))
cat(sprintf("BGS incip range %.0f-%.0f%%, EMS %.0f-%.0f%%\n",
            min(bgs), max(bgs), min(ems), max(ems)))

# ---------------- FIGURE (3 panels) ----------------
# matplotlib figsize (16.5, 4.9) at dpi 170 -> 2805 x 833 px
png(file.path(OUT_DIR, "CJFR-absolute-fig.png"),
    width = 2805, height = 833, res = 170, pointsize = 10)
par(mfrow = c(1, 3), mgp = c(2.4, 0.7, 0))
par(cex = 1)   # undo the automatic mfrow cex shrink so text matches matplotlib
OHM <- intToUtf8(0x03A9); MIDDOT <- intToUtf8(0x00B7)
EMDASH <- intToUtf8(0x2014); ELLIP <- intToUtf8(0x2026)
ohm_m <- paste0(OHM, MIDDOT, "m")
spp_order <- c("A.rubrum", "T.canadensis", "N.sylvatica", "Q.rubra")

# A: absolute resistivity by species (the confound made visible)
set.seed(0)
par(mar = c(4, 6.5, 3.5, 1))
plot(NA, xlim = rev(range(m$res) + c(-25, 25)), ylim = c(-0.5, 3.5),
     xlab = paste0("mean resistivity (", ohm_m, ")  ", EMDASH, " lower = wetter"),
     ylab = "", yaxt = "n", main = "")
for (i in seq_along(spp_order)) {
  s <- spp_order[i]
  d <- m$res[m$sp == s]
  points(d, rnorm(length(d), i - 1, 0.08), col = "white",
         bg = adjustcolor(SPECIES_COLS[[s]], alpha.f = 0.8), pch = 21, cex = 1.15, lwd = 0.5)
  segments(mean(d), i - 1 - 0.28, mean(d), i - 1 + 0.28, col = SPECIES_COLS[[s]], lwd = 2.5)
}
axis(2, at = 0:3, las = 1, cex.axis = 0.95,
     labels = paste0(spp_order, ifelse(spp_order %in% c("N.sylvatica", "Q.rubra"), "*", "")))
title(main = "A.  Absolute axis is species-confounded\n(* = single-site specialist)",
      adj = 0, cex.main = 1.05, font.main = 1)

# B: predicted-moisture distribution by site with candidate thresholds
par(mar = c(4, 4, 3.5, 1))
xs2 <- seq(min(m$pmoist) - 5, max(m$pmoist) + 5, length.out = 300)
kb <- approx(density(m$pmoist[mb])$x, density(m$pmoist[mb])$y, xs2, rule = 2)$y
ke <- approx(density(m$pmoist[me])$x, density(m$pmoist[me])$y, xs2, rule = 2)$y
ymax <- max(kb, ke) * 1.05
plot(NA, xlim = range(xs2), ylim = c(-0.0022, ymax),
     xlab = "predicted moisture content (%)  [from validation calibration]",
     ylab = "density", main = "")
lines(xs2, kb, col = "#1f77b4", lwd = 2)
lines(xs2, ke, col = "#d9822b", lwd = 2)
points(m$pmoist[mb], rep(-0.001, sum(mb)), col = adjustcolor("#1f77b4", 0.6), pch = 16, cex = 0.55)
points(m$pmoist[me], rep(-0.0006, sum(me)), col = adjustcolor("#d9822b", 0.6), pch = 16, cex = 0.55)
for (mc in list(c(90, "#999999"), c(100, "#555555"), c(110, "#111111"))) {
  x0 <- as.numeric(mc[[1]])
  abline(v = x0, col = mc[[2]], lty = 3, lwd = 1.2)
  text(x0, ymax * 0.92, sprintf("%d%%", x0), cex = 0.6)
}
legend("topright", legend = c("BGS", "EMS"), col = c("#1f77b4", "#d9822b"),
       lwd = 2, bty = "n", cex = 0.9)
title(main = "B.  Absolute (predicted moisture) axis by site",
      adj = 0, cex.main = 1.05, font.main = 1)

# C: site direction sweep on absolute resistivity, anchors marked
par(mar = c(4, 4, 3.5, 1))
ymax2 <- max(bgs, ems) * 1.04
plot(NA, xlim = rev(range(grid)), ylim = c(0, ymax2),
     xlab = paste0("anomaly threshold: resistivity below ", ELLIP, " (", ohm_m, ")"),
     ylab = "% of site trees \"incipient\"", main = "")
kp <- bgs >= ems
if (any(kp)) polygon(c(grid[kp], rev(grid[kp])), c(bgs[kp], rev(ems[kp])),
                     col = adjustcolor("#1f77b4", alpha.f = 0.10), border = NA)
ko <- ems > bgs
if (any(ko)) polygon(c(grid[ko], rev(grid[ko])), c(bgs[ko], rev(ems[ko])),
                     col = adjustcolor("#d9822b", alpha.f = 0.18), border = NA)
lines(grid, bgs, col = "#1f77b4", lwd = 2.4)
lines(grid, ems, col = "#d9822b", lwd = 2.4)
anchor <- c(median = median(res), mean = mean(res), Otsu = t_otsu,
            "moist>100%" = r100, "wetwood~200" = 200)
for (nm in names(anchor)) {
  abline(v = anchor[[nm]], col = "#444444", lty = 3, lwd = 1.0)
  text(anchor[[nm]], ymax2 * 0.98, nm, srt = 90, adj = c(1, 0), cex = 0.62, col = "#444444")
}
legend("topleft", legend = c("BGS (wetland)", "EMS (upland)"),
       col = c("#1f77b4", "#d9822b"), lwd = 2.4, bty = "n", cex = 0.9)
title(main = "C.  Site direction across absolute thresholds\n(shaded orange = where upland exceeds wetland)",
      adj = 0, cex.main = 1.05, font.main = 1)
invisible(dev.off())
cat("\nsaved figure + CSV\n")
